### ISSUE TITLE
Why not use early_stop_rounds?

### ISSUE THREAD

**Issue by stepthom** [microsoft/FLAML#172]
For LGBM and XGBoost,  `num_estimators` is sampled between 4 and `min(32768, int(data_size))`. 

Instead, have you considered setting `num_estimators` (alias `num_boost_round`) to a large value (say 32768) and using `evals` and `early_stop_rounds` during training? This would allow the learning algorithm (rather than the tuning algorithm) to directly find the number of boosting rounds that is optimal, given the data and all other hyper parameter values.  (The algorithms will output the best num_estimators/num_boost_round via `best_iteration`.)

And then, when you refit on the full dataset, you can use the (average/mode/etc) `best_iteration` that were used during CV.

**Comment by sonichi** [comment]
It's an idea worth trying. Would you like to implement and test it?

**Comment by stepthom** [comment]
@sonichi I have done some initial experimentation and it has led to accuracy improvement.👍 

I will not have time to robustly implement in FLAML for a while, but just wanted to bring to your attention

**Comment by qingyun-wu** [comment]
> @sonichi I have done some initial experimentation and it has led to accuracy improvement.👍
> 
> I will not have time to robustly implement in FLAML for a while, but just wanted to bring to your attention

Thanks, @stepthom for exploring this. The suggestion sounds interesting and useful. We (me, chi, or some other volunteers) can probably find time to work on it. I wonder could you please share with us your initial experimentation? Will also keep you posted about the progress in case you have comments. Thank you!

**Comment by sonichi** [comment]
@stepthom Any update?

**Comment by stepthom** [comment]
@sonichi I am working on this now, and will have something to post by the end of the week.

**Comment by stepthom** [comment]
Here are some initial results. You can find my code and results here: https://github.com/stepthom/vaccine.

I'm coping the contents of the README.md file here for convenience: 

# Early stopping experiment on vaccine dataset 
Experiment to test early stopping during hyperparameter tuning of LGBM in FLAML.  See related discussion in https://github.com/microsoft/FLAML/issues/172.

# Experiment Setup

- Data is from [DrivenData's vaccine challenge](https://www.drivendata.org/competitions/66/flu-shot-learning/).
- Data contains ~27K instances and 34 features.
- Data contains both categorical and numeric features. 
- Minimal preprocessing is applied to the data.
- I used FLAML's tune api.
- I used only LGBM.
- I tuned the same hyperparameters, using the same value ranges, as FLAML does in [model.py](https://github.com/microsoft/FLAML/blob/a99e939404caeda88f32724cc264841f2f5dcfca/flaml/model.py#L215).
- For each candidate set of hyperparameter values, I ran [stratified K-fold cross validation](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.StratifiedKFold.html) with `n_splits` set to 15 to get an accurate estimate of performance.


# Experiments

I created two experiments that share the setup above.

- Experiment 1 _did not_ use early stopping. `n_estimators` is sampled as part of the tuning process.
- Experiment 2 _did_ use early stopping. I set `n_estimators` to the upper bound (i.e., 32768). I set `early_stopping_rounds` to 100.

For each experiment, I tuned with a time budget of roughly 8 hours (30000 seconds).


# Results

Detailed results can be found in [analyze_results.ipynb](https://github.com/stepthom/vaccine/blob/main/analyze_results.ipynb). 

Some initial takeawars are as follws. Early stopping: 
- allowed more iterations/trials to be completed in the same amount of time (799 vs 192)
- found a slightly better cross validation score (0.8605 vs 0.8596)
- found the best trial faster (after 4104 seconds compared to 6621 seconds)

**Comment by sonichi** [comment]
Thanks for the initial results! Quick comment:
```
    low_cost_partial_config={
        'num_leaves': 4,
        'min_child_samples': 20,
        'learning_rate': 1.0,
        'max_bin': 8,
        'colsample_bytree': 0.1,
        "reg_alpha": 1/1024,
        "reg_lambda": 1.0,
    }
```
can be simplified to `low_cost_partial_config={'num_leaves': 4}`, which might give you better results.

**Comment by sonichi** [comment]
@stepthom I tried adding early_stopping_rounds in the search space. The results are not conclusive. In some cases the results are better, and in other cases they are worse. My change is in the `early_stop` branch. If you'd like to improve it, I can help with benchmarking.
If you are satisfied with using the tune api for lgbm with early stopping, then we can close this issue.

**Comment by stepthom** [comment]
@sonichi. Very interesting. Thanks for looking into it. We can close the issue; I am happy with using tune api directly for now. I will continue to do research some research as to how/why/when early_stopping would work and when it wouldn't. Out of curiosity, does FLAML have any "benchmark" datasets and results to which we can compare results to know if a given approach is beneficial?

**Comment by sonichi** [comment]
@stepthom Thanks. Yes, there is an [automlbenchmark](https://github.com/openml/automlbenchmark) based on which we can evaluate each new version and compare with the released version. Looking forward to your research and improvement of the library.
